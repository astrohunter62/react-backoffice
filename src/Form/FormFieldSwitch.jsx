import React from 'react'
import PropTypes from 'prop-types'

import {
  FormControlLabel,
  FormHelperText,
  Switch,
} from '@material-ui/core'

class FormFieldSwitch extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    handleChange: PropTypes.func.isRequired,
    helperText: PropTypes.string,
    className: PropTypes.string,
    isDisabled: PropTypes.bool,
  }

  static defaultProps = {
    helperText: null,
    className: '',
    isDisabled: false,
  }

  state = {
    value: false,
  }

  handleChange(id) {
    const func = this.props.handleChange(id)

    return () => {
      const newValue = !this.state.value

      this.setState({
        value: newValue,
      })

      return func({
        target: {
          value: newValue,
        },
      })
    }
  }

  render() {
    const {
      id,
      title,
      helperText,
      isDisabled,
      className,
    } = this.props

    return (
      <div className={className}>
        <FormControlLabel
          disabled={isDisabled}
          control={
            <Switch
              checked={this.state.value}
              onChange={this.handleChange(id)}
              value={id}
              color="primary"
            />
          }
          label={title}
        />

        {helperText ? (
          <FormHelperText
            disabled={isDisabled}
          >
            {helperText}
          </FormHelperText>
        ) : null}
      </div>
    )
  }
}

export default FormFieldSwitch
