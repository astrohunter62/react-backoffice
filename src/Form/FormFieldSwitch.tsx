import React from "react";
import { FormControlLabel, FormHelperText, Switch } from "@material-ui/core";

type FormFieldSwitchProps = {
  id: string;
  title: string;
  handleChange: (...args: any[]) => any;
  helperText?: string;
  className?: string;
  isDisabled?: boolean;
};

type FormFieldSwitchState = {
  value: boolean;
};

class FormFieldSwitch extends React.Component<
  FormFieldSwitchProps,
  FormFieldSwitchState
> {
  state = {
    value: false,
  };

  handleChange(id: string) {
    const { handleChange } = this.props;
    const { value } = this.state;
    const func = handleChange(id);
    return () => {
      const newValue = !value;
      this.setState({
        value: newValue,
      });
      return func({
        target: {
          value: newValue,
        },
      });
    };
  }

  render() {
    const { id, title, helperText, isDisabled, className } = this.props;
    const { value } = this.state;
    return (
      <div className={className}>
        <FormControlLabel
          disabled={isDisabled}
          control={
            <Switch
              checked={value}
              onChange={this.handleChange(id)}
              value={id}
              color="primary"
            />
          }
          label={title}
        />

        {helperText ? (
          <FormHelperText disabled={isDisabled}>{helperText}</FormHelperText>
        ) : null}
      </div>
    );
  }
}

export default FormFieldSwitch;
