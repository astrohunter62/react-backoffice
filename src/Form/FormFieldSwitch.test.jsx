import React from 'react'
import renderer from 'react-test-renderer'

import FormFieldSwitch from './FormFieldSwitch'

it('renders correctly', () => {
  const tree = renderer
    .create((
      <FormFieldSwitch
        id="switch"
        title="Title"
        handleChange={() => () => {}}
      />
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders disabled', () => {
  const tree = renderer
    .create((
      <FormFieldSwitch
        id="switch"
        title="Title"
        handleChange={() => () => { }}
        isDisabled
      />
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders with helper text', () => {
  const tree = renderer
    .create((
      <FormFieldSwitch
        id="switch"
        title="Title"
        handleChange={() => () => { }}
        helperText="test"
      />
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})
