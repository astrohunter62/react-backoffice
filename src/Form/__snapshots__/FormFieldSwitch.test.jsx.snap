// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className=""
>
  <label
    className="MuiFormControlLabel-root-1"
  >
    <span
      className="MuiSwitch-root-4"
    >
      <span
        className="MuiButtonBase-root-24 MuiIconButton-root-18 MuiSwitchBase-root-13 MuiSwitchBase-default-15 MuiSwitch-default-8"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={undefined}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-23"
        >
          <span
            className="MuiSwitch-icon-6"
          />
          <input
            checked={false}
            className="MuiSwitchBase-input-14"
            disabled={false}
            id={undefined}
            name={undefined}
            onChange={[Function]}
            tabIndex={undefined}
            type="checkbox"
            value="switch"
          />
        </span>
        <span
          className="MuiTouchRipple-root-26"
        />
      </span>
      <span
        className="MuiSwitch-bar-5"
      />
    </span>
    <span
      className="MuiTypography-root-33 MuiTypography-body1-42 MuiFormControlLabel-label-3"
    >
      Title
    </span>
  </label>
</div>
`;

exports[`renders disabled 1`] = `
<div
  className=""
>
  <label
    className="MuiFormControlLabel-root-1 MuiFormControlLabel-disabled-2"
  >
    <span
      className="MuiSwitch-root-4"
    >
      <span
        className="MuiButtonBase-root-24 MuiButtonBase-disabled-25 MuiIconButton-root-18 MuiIconButton-disabled-22 MuiSwitchBase-root-13 MuiSwitchBase-default-15 MuiSwitch-default-8 MuiSwitchBase-disabled-17 MuiSwitch-disabled-12"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={undefined}
        tabIndex={-1}
      >
        <span
          className="MuiIconButton-label-23"
        >
          <span
            className="MuiSwitch-icon-6"
          />
          <input
            checked={false}
            className="MuiSwitchBase-input-14"
            disabled={true}
            id={undefined}
            name={undefined}
            onChange={[Function]}
            tabIndex={undefined}
            type="checkbox"
            value="switch"
          />
        </span>
      </span>
      <span
        className="MuiSwitch-bar-5"
      />
    </span>
    <span
      className="MuiTypography-root-33 MuiTypography-body1-42 MuiFormControlLabel-label-3"
    >
      Title
    </span>
  </label>
</div>
`;

exports[`renders with helper text 1`] = `
<div
  className=""
>
  <label
    className="MuiFormControlLabel-root-1"
  >
    <span
      className="MuiSwitch-root-4"
    >
      <span
        className="MuiButtonBase-root-24 MuiIconButton-root-18 MuiSwitchBase-root-13 MuiSwitchBase-default-15 MuiSwitch-default-8"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={undefined}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-23"
        >
          <span
            className="MuiSwitch-icon-6"
          />
          <input
            checked={false}
            className="MuiSwitchBase-input-14"
            disabled={false}
            id={undefined}
            name={undefined}
            onChange={[Function]}
            tabIndex={undefined}
            type="checkbox"
            value="switch"
          />
        </span>
        <span
          className="MuiTouchRipple-root-26"
        />
      </span>
      <span
        className="MuiSwitch-bar-5"
      />
    </span>
    <span
      className="MuiTypography-root-33 MuiTypography-body1-42 MuiFormControlLabel-label-3"
    >
      Title
    </span>
  </label>
  <p
    className="MuiFormHelperText-root-57"
  >
    test
  </p>
</div>
`;
